## Data Analysis Process

- The data analysis process is done in the following steps
    1. Asking Right Question About Data
    1. Data Wrangling/Preprocessing/Munging
        - Gathering Data
            - CSV files
            - API
            - Database
            - Web Scrapping etc.
        - Assessing Data
        - Cleaning Data 
    1. Exploratory Data Analysis(EDA) 
    1. Drawing Conclusion
    1. Communicating Results/Data Storytelling

In [2]:
import pandas as pd

## Gathering Data and Importing it

In [3]:
patient_df = pd.read_csv("./dataset/patients.csv")
patient_df.head(1)

patient_id assigned_sex given_name  surname               address  \
0           1       female        Zoe  Wellish  576 Brown Bear Drive   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   

                                contact  birthdate  weight  height   bmi  
0  951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6

In [4]:
treatment_df = pd.read_csv("./dataset/treatments.csv")
treatment_df.head(1)

given_name   surname    auralin novodra  hba1c_start  hba1c_end  \
0   veronika  jindrová  41u - 48u       -         7.63        7.2   

   hba1c_change  
0           NaN

In [5]:
adv_reaction_df = pd.read_csv("./dataset/adverse_reactions.csv")
adv_reaction_df.head(1)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort

In [6]:
patient_df.shape, treatment_df.shape, adv_reaction_df.shape

((503, 14), (280, 7), (34, 3))

## Assessing Data

It is about analysing and assesment of data so we can clean it in the next step. 

### Types of Unclean data
There are 2 kinds of unclean data
- Dirty Data (Data with Quality/Content Issue) 
- Messy Data (Data with tidiness/structural Issue)

1. **Dirty Data**
- It has following kind of issues
    - Duplicate Data
    - Missing Data
    - Corrupt Data
    - Inaccurate Data 

1. **Messy Data**
- Tidy data has following traits
    - Each Variable forms a column
    - Each observation forms a row
    - Each Observational unit forms a table

- If any of the above trait is voilated then it is untidy/messy data.

### Type of Assessment

There are 2 types of assessment styles

- Manual: Looking through data manually in google sheets
- Programmatic: By using pandas function like info(), describe(), sample(), etc.

Steps in Assessment include 2 steps:
- Discover
- Document

We first discover the issues and then document it and later in the data cleaning step all these issues are resolved.

### Data Quality Dimensions

There are 4 major type of quality issues
1. **Completeness Issues**: Data is missing
    - To handle missing data we can do imputaion or dropping.
2. **Validity Issues**: Data is available but not valid
3. **Accuracy Issues**: Data is available and valid but not accuratae
4. **Consistency Issues**: Data is inconsistent


In [11]:
patient_df.info(), patient_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


(None,
        patient_id      zip_code      weight      height         bmi
 count  503.000000    491.000000  503.000000  503.000000  503.000000
 mean   252.000000  49084.118126  173.434990   66.634195   27.483897
 std    145.347859  30265.807442   33.916741    4.411297    5.276438
 min      1.000000   1002.000000   48.800000   27.000000   17.100000
 25%    126.500000  21920.500000  149.300000   63.000000   23.300000
 50%    252.000000  48057.000000  175.300000   67.000000   27.200000
 75%    377.500000  75679.000000  199.500000   70.000000   31.750000
 max    503.000000  99701.000000  255.900000   79.000000   37.700000)

In [12]:
treatment_df.info(), treatment_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


(None,
        hba1c_start   hba1c_end  hba1c_change
 count   280.000000  280.000000    171.000000
 mean      7.985929    7.589286      0.546023
 std       0.568638    0.569672      0.279555
 min       7.500000    7.010000      0.200000
 25%       7.660000    7.270000      0.340000
 50%       7.800000    7.420000      0.380000
 75%       7.970000    7.570000      0.920000
 max       9.950000    9.580000      0.990000)

In [13]:
adv_reaction_df.info(), adv_reaction_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


(None,
        given_name  surname adverse_reaction
 count          34       34               34
 unique         34       33                6
 top         berta  johnson     hypoglycemia
 freq            1        2               19)

### Quality Issues in the DataFrames (Discovering and Documenting)

- Data Quality Dimension: [Completeness, Validity, Accuracy, Consistency]

#### `Patients DF`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[A]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments DF`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[A]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions DF`

- Name and surname col is in lower case[IC]

### Tidiness Issues in the DataFrames (Discovering and Documenting)

#### `Patients DF`

- Phone and email together in one col i.e. contact

#### `Treatments DF`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions DF`

- No point of using this table

## Cleaning Data

While cleaning you should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define what you need to clean
- Code to clean
- Test if data is cleaned properly or not

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

### Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

### Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table